tensorflow/core/kernels/conv_ops_gpu_3.cu.cc

```diff
@@ -422,29 +595,131 @@ void RunSwapDimension1And2InTensor3(const GPUDevice& d, const T* input,
   // If both dimensions are not trivial, use tiles for the actual swapping.
   // Otherwise, the trivial swapping relying on the ldg cache is more efficient.
   static const int kMinDimensionToUseTiles = 16;
-  bool use_tiles = (input_dims[1] >= kMinDimensionToUseTiles &&
-                    input_dims[2] >= kMinDimensionToUseTiles);
-  if (use_tiles) {
+  static const int kMinDimensionToUseRectTiles = 96;
+
+  bool large_matrix = (input_dims[1] >= kMinDimensionToUseTiles &&
+                       input_dims[2] >= kMinDimensionToUseTiles);
+  bool narrow_matrix = (input_dims[1] >= kMinDimensionToUseRectTiles ||
+                      input_dims[2] >= kMinDimensionToUseRectTiles);
+  if (large_matrix) {
     // We get best performance when TileSize is the number of threads in a warp
     // (32 on our GPUs) and NumSubTiles is 8, so our block size is 8 * 32 = 256
     // threads.
     static const int TileSize = 32;
-    static const int NumSubTiles = 8;
+    static const int ThreadNum = 256;
+
     Dimension<3> input_dims_in_tiles = {
         input_dims[0], (input_dims[1] + TileSize - 1) / TileSize,
         (input_dims[2] + TileSize - 1) / TileSize,
     };
+
+    int total_tiles_count = input_dims_in_tiles[0] * input_dims_in_tiles[1] *
+                            input_dims_in_tiles[2];
+    SwapDimension1And2InTensor3UsingTiles<
+        T, ThreadNum, TileSize,
+        TileSize><<<total_tiles_count, ThreadNum, 0, d.stream()>>>(
+        input, input_dims, output);
+
+  } else if (narrow_matrix) {
+    // Define available tile sizes here for each size of data type supported:
+    std::map<int, int> tile_spec_128 = {{32, 15}, {64, 15}, {128, 15},
+                                        {256, 2}};
+    std::map<int, int> tile_spec_64  = {{32, 15},  {64, 15}, {128, 15},
+                                        {256, 8},  {512, 2}};
+    std::map<int, int> tile_spec_32  = {{32, 15},  {64, 15}, {128, 15},
+                                        {256, 10}, {512, 4}, {1024, 2}};
+    std::map<int, int> tile_spec_16  = {{32, 15},  {64, 15}, {128, 15},
+                                        {256, 10}, {512, 4}, {1024, 2}};
+    std::map<int, int> tile_spec_8   = {{32, 15},  {64, 15}, {128, 15},
+                                        {256, 10}, {512, 4}, {1024, 2}};
```

Comment:
I'm concerned constructing all these std::maps may add a nontrivial overhead to kernel launches.  std::map is kind of evil.

As a first change, it seems like you could accomplish basically the same thing with fewer maps by doing something like:

```
  std::map<int, int> tile_spec = []() -> std::map<int, int> {
  switch (sizeof(T)) {
    case 1: return {{32, 15}, ...};
    case 2: ...
  }();
```

But looking closer, it seems that we also don't use the fact that this is a map, as opposed to, say, a `vector<pair<int, int>>`?  We could optimize further, but constructing one small vector is probably fine in terms of overhead.

What do you think of pulling this whole computation out into a separate function?